### PR TITLE
SAM-1074 Adding Delete test submission attempt feature for samigo, we…

### PR DESCRIPTION
**Pull request by yuanhuaqu**
… used this feature at Texas State Univ  from 2006 till now for sakai10.4

Tested over sakai's trunk, works well. For history of this feature merge request and also related feature demo video link, please see SAM-1074.

### CODE REVIEW

**Comment by bjones86** [comment]
Can you amend the last commit to prefix it with SAM-1074 please?

File: samigo/samigo-app/src/webapp/jsf/evaluation/totalScores.jsp [hunk 1]
```diff
@@ -87,6 +87,21 @@ function disableIt()
...
+			this.onclick = null;
+			$(this).click(function(){
+				if ( confirm("Are you sure you want to delete this attempt?") ) {
+					this.existingOnclick();
+				} else {
+					return false;
+				}
+			});
+		});
+	});
```
Comment (bjones86) [inline samigo/samigo-app/src/webapp/jsf/evaluation/totalScores.jsp]: Tabs vs. spaces here looks inconsistent with the rest of the file

File: samigo/samigo-services/src/java/org/sakaiproject/tool/assessment/facade/AssessmentGradingFacadeQueries.java [hunk 2]
```diff
@@ -223,6 +224,7 @@ public Object doInHibernate(Session session) throws HibernateException, SQLExcep
         return list;
       }
       else {
+        if(list.isEmpty()) return new ArrayList();
```
Comment (bjones86) [inline samigo/samigo-services/src/java/org/sakaiproject/tool/assessment/facade/AssessmentGradingFacadeQueries.java]: Braces on one line if/else statements please

File: samigo/samigo-services/src/java/org/sakaiproject/tool/assessment/integration/helper/integrated/GradebookServiceHelperImpl.java [hunk 3]
```diff
@@ -255,22 +255,37 @@ public void updateExternalAssessmentScore(AssessmentGradingData ag,
...
     if (gradebookUId == null)
     {
       return;
     }
     
-    //SAM-1562 We need to round the double score and covert to a double -DH
-    double fScore = MathUtils.round(ag.getFinalScore(), 2);
-    Double score = Double.valueOf(fScore).doubleValue();
-    log.info("rounded:  " + ag.getFinalScore() + " to: " + score.toString() );
+    //Will pass to null value when last submission is deleted -Qu 1/7/2013
```
Comment (bjones86) [inline samigo/samigo-services/src/java/org/sakaiproject/tool/assessment/integration/helper/integrated/GradebookServiceHelperImpl.java]: Name/dates not necessary in code comments

File: samigo/samigo-services/src/java/org/sakaiproject/tool/assessment/services/GradingService.java [hunk 4]
```diff
@@ -278,6 +278,48 @@ public void saveTotalScores(ArrayList gdataList, PublishedAssessmentIfc pub)
...
+   * @param gdataList  a list of AssessmentGradingData
+   * @param pub   PublishedAssessment object
+   * @param studentId
+   */
+
+  public void notifyDeleteToGradebook(ArrayList gdataList, PublishedAssessmentIfc pub, String studentId){
+    try {
+      AssessmentGradingData gdata = null;
+      if (gdataList.size()>0)
+        gdata = (AssessmentGradingData) gdataList.get(0);
```
Comment (bjones86) [inline samigo/samigo-services/src/java/org/sakaiproject/tool/assessment/services/GradingService.java]: Braces on one line if/else statements. Same with below

File: samigo/samigo-services/src/java/org/sakaiproject/tool/assessment/services/GradingService.java [hunk 5]
```diff
@@ -2048,6 +2090,16 @@ public ArrayList getLastItemGradingDataPosition(Long assessmentGradingId, String
 	    return results;
 	  }
   
+  public List getItemGradingIds(Long assessmentGradingId) {
+		List results = null;
+	    try {
+			results = PersistenceService.getInstance().getAssessmentGradingFacadeQueries().getItemGradingIds(assessmentGradingId);
```
Comment (bjones86) [inline samigo/samigo-services/src/java/org/sakaiproject/tool/assessment/services/GradingService.java]: Spacing looks inconsistent here

**Comment by bjones86** [comment]
Code comments provided in-line

**Comment by ern** [comment]
This feature should have a property to control whether it can be turned on, and should be off by default.

**Comment by jonespm** [comment]
This jira has merge conflict and without it, the feature on SAM-2587 doesn't work and in turn https://jira.sakaiproject.org/browse/SAM-2712 results.

Are @yuanhuaqu able to resolve the conflicts here so this can be merged? The property is probably a good idea but would need to involve both this workflow and another and I think should be a separate issue.

**Comment by jonespm** [comment]
This looks good, can you run a "git rebase master" to clean up all the commits on this?

**Comment by yuanhuaqu** [comment]
I did a merge with master on Mar 22 and resolved the conflicts.  I tried to do "git rebase master" , it somehow didn't pull all the commit on this branch and allow me to squash.  I guess, having a merge in the middle of the process won't allow me to rebase? Any guide I can follow for this case?

**Comment by ern** [comment]
Hi @yuanhuaqu 
Having a merge in there does complicate things a little..
Basically you need to get rid of the merge, you can do this a few ways perform a rebase and drop the commit with the merge.You can also squash your commits in the same rebase.
So you need to figure out how many commits to include in the rebase, in the example below I use 5 but you should adjust according to the number of commits you have on this branch. (Remember the one of those commits is the merge)
git rebase HEAD~5
Select which commits to squash and select the merge commit to drop.
Once this is complete you will then just have 1 commit and the merge will be gone.
Now in order to update your branch from master you perform from your branch:
git rebase master
This will then bring your branch up to date with master (you may need to fix if there are conflicts).
At this point you should have just 1 commit being the last one, now just need to update this PR by force pushing your changes.
Hope this helps!
BTW there are others to do this like performing a SOFT reset on your branch but the way I described above would be the most direct.

**Comment by yuanhuaqu** [comment]
It took me a while figuring out the complexity caused by using merge instead of rebase for updating  my feature branch with master, resolve conflicts and eventually squash all the commits (8) into a cleaner one.

@ern  
Thanks  for the guide.  It helped me a great deal!!!!    May I suggest you to put this guide in somewhere like [Git+Setup](https://confluence.sakaiproject.org/display/SAKDEV/Git+Setup ) ?  And surely, case like mine with this branch which almost stretch out about a year,   to keep feature branch updated before it finally get into the master base , doing a   `git rebase master` will be much simpler than `git merge master` (which I learned it the hard way :=))

**Comment by bjones86** [comment]
I have some internal documentation explaining most if not all of this, which I plan to move into the Sakai Confluence page you mentioned in the near future.

**Comment by jonespm** [comment]
Thanks all for following up on this, I'll merge and test this after Travis passes.

**Comment by yuanhuaqu** [comment]
Thanks.

**Pull request merged**